nntrainer/dataset/random_data_producers.cpp

```diff
@@ -169,4 +171,74 @@ RandomDataOneHotProducer::finalize(const std::vector<TensorDim> &input_dims,
     return DataProducer::Iteration(false, inputs, labels);
   };
 }
+
+DataProducer::Generator_sample RandomDataOneHotProducer::finalize_sample(
+  const std::vector<TensorDim> &input_dims,
+  const std::vector<TensorDim> &label_dims) {
+  /** check if the given producer is ready to finalize */
+  nntrainer::PropsMin min_;
+  nntrainer::PropsMax max_;
+  std::tie(min_, max_, std::ignore) = *rd_one_hot_props;
+
+  /// @todo expand this to non onehot case
+  NNTR_THROW_IF(std::any_of(label_dims.begin(), label_dims.end(),
+                            [](const TensorDim &dim) {
+                              return dim.channel() != 1 || dim.height() != 1;
+                            }),
+                std::invalid_argument)
+    << "Label dimension containing channel or height not allowed";
+
+  NNTR_THROW_IF(min_.get() > max_.get(), std::invalid_argument)
+    << "Min value is bigger then max value, min: " << min_.get()
+    << "max: " << max_.get();
+
+  /// @todo move this to higher order component
+  NNTR_THROW_IF(size(input_dims, label_dims) == 0, std::invalid_argument)
+    << "size is zero, dataproducer does not provide anything";
+
+  /** prepare states for the generator */
+  std::vector<std::uniform_int_distribution<unsigned int>> label_chooser_;
+  label_chooser_.reserve(label_dims.size());
+  std::transform(label_dims.begin(), label_dims.end(),
+                 std::back_inserter(label_chooser_),
+                 [this](const TensorDim &label_dim) {
+                   return std::uniform_int_distribution<unsigned int>(
+                     0, label_dim.width() - 1);
+                 });
+
+  std::mt19937 rng;
+  rng.seed(getSeed());
+  auto sz = size(input_dims, input_dims);
+
+  /** DataProducer::Generator_sample */
+  return [rng, sz, input_dims, label_dims, min_ = min_.get(), max_ = max_.get(),
+          current_iteration = 0ULL, label_chooser = std::move(label_chooser_)](
+           unsigned int idx, DataProducer::Sample &sample) mutable -> bool {
+    auto populate_input = [&](Tensor *t) {
+      t->setRandUniform(min_, max_);
+      return t;
+    };
+
+    auto populate_label =
+      [&](Tensor *t, std::uniform_int_distribution<unsigned int> &label_dist_) {
+        t->setZero();
+        t->setValue(0, 0, 0, label_dist_(rng), 1);
+        return t;
+      };
+
+    auto &[inputs, labels] = sample;
+    std::transform(inputs.begin(), inputs.end(), inputs.begin(),
```

Comment:
I recommend to use `std::for_each` here as you intend to perform the function on each element in-place. WIth this you dont need to return in your lambdas and it also simplifies the usage.

Reply:
Done